File: 2020/5.2.rb
require_relative '5.1.rb'

module DayFivePartTwo
  module_function

  def solve(file)
    inputs = DayFivePartOne.prepare_inputs(file)
    ids = get_ids(inputs)
    find_missing_id(ids)
  end

  def find_missing_id(ids)
    (0..1016).find do |num|
      puts num
      !ids.count(num - 1).zero? &&
        !ids.count(num + 1).zero?
    end
  end

  def get_ids(inputs)
    inputs.map do |input|
      row_index_end = 127
      row_index_start = 0
      column_index_start = 0
      column_index_end = 7
      input.split('').each do |char|
        if char == 'B'
          row_index_start = row_index_end - (row_index_end - row_index_start) / 2
        elsif char == 'F'
          row_index_end -= (row_index_end - row_index_start) / 2
        elsif char == 'R'
          column_index_start = column_index_end - (column_index_end - column_index_start) / 2
        else
          column_index_end -= (column_index_end - column_index_start) / 2
        end
      end
      row_index_start * 8 + column_index_start
    end
  end
end

# puts DayFivePartTwo.solve(__dir__ + '/5test.txt')
puts DayFivePartTwo.solve(__dir__ + '/5.txt')

File: 2021/05/5.2.php
<?php


require "5.1.php";
class DayFivePartTwo
{
    function solve($filename)
    {
        $inputs = DayFivePartOne::prepareInputs($filename);
        $grid = DayFivePartOne::fillGrid($inputs, $this->filterLines($inputs));
        echo DayFivePartOne::calculateTotalOverlaps($grid, 2) . "\n";
    }

    static function filterLines($inputs)
    {
        return array_reduce($inputs, function ($acc, $input) {
            if ($input[0][0] ==  $input[1][0] ||  $input[0][1] ==   $input[1][1]) {
                $acc = array_merge([["inputs" => $input, "diagonal" => false]], $acc);
            } else if (abs($input[0][1] - $input[1][1]) == abs($input[0][0] - $input[1][0])) {
                $acc = array_merge([["inputs" => $input, "diagonal" => true]], $acc);
            }
            return $acc;
        }, []);
    }
}

$class = new DayFivePartTwo();
// $class->solve("/2021/5test.txt");
$class->solve("/2021/5.txt");
